h/static/scripts/annotator/anchoring/html.coffee

```diff
@@ -0,0 +1,98 @@
+raf = require('raf')
+
+{
+  FragmentAnchor
+  RangeAnchor
+  TextPositionAnchor
+  TextQuoteAnchor
+} = require('./types')
+
+
+querySelector = (type, root, selector, options) ->
+  doQuery = (resolve, reject) ->
+    try
+      anchor = type.fromSelector(root, selector, options)
+      range = anchor.toRange(options)
+      resolve(range)
+    catch error
+      reject(error)
+  return new Promise(doQuery)
+
+
+###*
+# Anchor a set of selectors.
+#
+# This function converts a set of selectors into a document range using.
+# It encapsulates the core anchoring algorithm, using the selectors alone or
+# in combination to establish the best anchor within the document.
+#
+# :param Element root: The root element of the anchoring context.
+# :param Array selectors: The selectors to try.
+# :param Object options: Options to pass to the anchor implementations.
+# :return: A Promise that resolves to a Range on success.
+# :rtype: Promise
+####
+exports.anchor = (root, selectors, options = {}) ->
+  # Selectors
+  fragment = null
+  position = null
+  quote = null
+  range = null
+
+  # Collect all the selectors
+  for selector in selectors ? []
+    switch selector.type
+      when 'FragmentSelector'
+        fragment = selector
+      when 'TextPositionSelector'
+        position = selector
+        options.hint = position.start  # TextQuoteAnchor hint
+      when 'TextQuoteSelector'
+        quote = selector
+      when 'RangeSelector'
+        range = selector
+
+  # Assert the quote matches the stored quote, if applicable
+  maybeAssertQuote = (range) ->
+    if quote?.exact? and range.toString() != quote.exact
+      throw new Error('quote mismatch')
+    else
+      return range
+
+  # Until we successfully anchor, we fail.
+  promise = Promise.reject('unable to anchor')
```

Comment:
When reading this pattern:

```coffeescript
promise = Promise.reject('boom')

if foo?
  promise = promise.catch ->
    return myNewPromise()

if bar?
  promise = promise.catch ->
    return myOtherPromise()

return promise
```

My first thought was: "would this be simpler and clearer as..."

```coffeescript
promise = null

if foo?
  promise = myNewPromise()

if bar?
  promise = myOtherPromise()

if not promise?
  return Promise.reject('boom')
return promise
```

Of course the answer is "no", because the entire point is that you're chaining up the anchoring strategies so that if one of them fails to anchor the next one is tried.

I think adding a comment to this effect here, and pointing out that the chain goes, top to bottom, from "fast but brittle" to "slow but robust", would be useful.